Implement Telemetry Metrics for Metadata Server

muhamadazmy opened the pull request:
Implement Telemetry Metrics for Metadata Server

Summary:
This pull request introduces telemetry metrics to the metadata server, enhancing monitoring and performance analysis capabilities.

Fixes #2605

muhamadazmy commented:
@tillrohrmann feel free to suggest more metrics to add that you think can be useful.

tillrohrmann reviewed (approved):
Thanks for adding metrics to the `RaftMetadataStore` @muhamadazmy. The changes look good to me. What we might also add as metrics are the applied index, committed index, snapshot size, trim index, leader, term, etc. (basically what we report in `MetadataStoreSummary`). This could be done as a follow-up if you want to merge this PR.

tillrohrmann commented:
It would be great to investigate why the e2e test is failing. I don't expect that your changes are causing it but we need to harden these test failures and make sure that it is not pointing towards a real problem.

muhamadazmy commented:
@tillrohrmann Great! thank you for your review. I wanted to add some of those metrics but wanted to make sure they would be useful since they already reported in the summary.

I will investigate why the tests are failing.

muhamadazmy commented:
Thank you @tillrohrmann for your review. I added more metrics as you suggested. I also added few simple metrics to the metadata store client as well as Ahmed suggested.

AhmedSoliman commented on crates/core/src/metadata_store.rs:
```diff
@@ -299,28 +306,43 @@ impl MetadataStoreClient {
     ) -> Result<Option<T>, ReadError> {
-        let value = self.inner.get(key).await?;
+        let start_time = Instant::now();
+        counter!(METADATA_CLIENT_GET_TOTAL, "key" => key.to_string()).increment(1);
```
I'm a little worried about the cardinality of this counter. We have a key per partition for epoch metadata.

muhamadazmy replied:
Yeah, i wasn't sure about this tbh. On other hand it also can help debugging irregular access pattern to a certain key.

What about if the key is prefixed with `pp_epoch` we only use that prefix for the metrics?

AhmedSoliman replied:
it's okay, let's leave the key for counters, but histogram per key is not great.

AhmedSoliman commented on crates/core/src/metadata_store.rs:
```diff
@@ -26,6 +32,7 @@ use restate_types::storage::{StorageCodec, StorageDecode, StorageEncode, Storage
 use restate_types::{flexbuffers_storage_encode_decode, Version, Versioned};
 use std::future::Future;
 use std::sync::Arc;
+use std::time::Instant;
```
In general, it's better to use tokio's Instant, but that's not an issue in this case. Just a reminder to form the habit.

AhmedSoliman commented on crates/core/src/metadata_store.rs:
```diff
@@ -299,28 +306,43 @@ impl MetadataStoreClient {
     pub async fn get_version(&self, key: ByteString) -> Result<Option<Version>, ReadError> {
-        self.inner.get_version(key).await
+        let start_time = Instant::now();
+        counter!(METADATA_CLIENT_GET_VERSION_TOTAL, "key" => key.to_string()).increment(1);
+        let result = self.inner.get_version(key.clone()).await;
+
+        histogram!(METADATA_CLIENT_GET_VERSION_DURATION, "key" => key.to_string())
+            .record(start_time.elapsed());
```
That's a lot of histograms (cardinality is a problem here as well)

muhamadazmy replied:
Yeah, Indeed. How do you feel about the suggested solution above. Another solution of course it to drop the key label completely

AhmedSoliman commented on crates/metadata-server/src/grpc/handler.rs:
```diff
@@ -53,6 +60,9 @@ impl MetadataStoreHandler {
 #[async_trait]
 impl MetadataServerSvc for MetadataStoreHandler {
     async fn get(&self, request: Request<GetRequest>) -> Result<Response<GetResponse>, Status> {
+        counter!(METADATA_SERVER_GET_TOTAL).increment(1);
```
it would have been nice if the counter has a dimension for status (success vs. failure). Imagine wanting to monitor success rate of a give metadata server, how would we do that?

This comment applies to all fallible operations.

muhamadazmy replied:
Yeah, that's definitely a good idea. I will update it

AhmedSoliman commented on crates/metadata-server/src/metric_definitions.rs:
```diff
@@ -0,0 +1,159 @@
+
+// Raft specific metrics
+pub(crate) const METADATA_SERVER_RAFT_SENT_MESSAGE_TOTAL: &str =
+    "restate.metadata_server.raft.sent_messages.total";
```
Do we need the `raft` part in this?

muhamadazmy commented:
Thanks @AhmedSoliman for your review. I have processed all your comments. I also dropped the `key` label from the client metrics, and recorded separate counter for both success and error for both client and server.

AhmedSoliman commented:
Do we have per-key histograms on the server metrics?

AhmedSoliman commented on crates/core/src/metadata_store.rs:
```diff
@@ -299,28 +325,42 @@ impl MetadataStoreClient {
-            Ok(None)
-        }
+        metrics_helper!(
+            @success=METADATA_CLIENT_GET_TOTAL_SUCCESS,
+            @error=METADATA_CLIENT_GET_TOTAL_ERROR,
+            @duration=METADATA_CLIENT_GET_DURATION,
```
Separating successes and errors into different metric names makes hard at query time to aggregate. Maybe take a look at an example in task_center.rs that emits the status as dimension instead.
```rust
counter!(TC_FINISHED, "kind" => kind_str, "status" => TC_STATUS_FAILED)
    .increment(1);
```
or from ingress:
```rust
counter!(
    INGRESS_REQUESTS,
    "status" => REQUEST_COMPLETED,
    "rpc.service" => service_name,
    "rpc.method" => handler_name,
)
.increment(1);
````

muhamadazmy replied:
You are right! I should have thought of that :D

AhmedSoliman commented on crates/core/src/metadata_store.rs:
```diff
@@ -263,6 +274,21 @@ impl<T: ProvisionedMetadataStore + Sync> MetadataStore for T {
     }
 }
 
+macro_rules! metrics_helper {
+    (@success=$ok:ident, @error=$err:ident, @duration=$dur:ident, $block:block) => {{
+        let start_time = tokio::time::Instant::now();
+        let result = $block;
+        metrics::histogram!($dur).record(start_time.elapsed());
+        if result.is_ok() {
+            metrics::counter!($ok).increment(1);
+        } else {
+            metrics::counter!($err).increment(1);
+        }
+
+        result
+    }};
+}
+
```
I'm not sure if it's worth a macro if this is used only a couple of times to be honest. I'm not sure if it _needs_ a macro if it's just the dimension name that will report the status.

muhamadazmy replied:
It's only to not repeat myself. It's easier to make changes for example.

AhmedSoliman commented on crates/metadata-server/src/metric_definitions.rs:
```diff
@@ -0,0 +1,166 @@
+    describe_gauge!(
+        METADATA_SERVER_EMBEDDED_TERM,
+        Unit::Count,
+        "Raft Metadata raft term number"
+    );
+
+    describe_gauge!(
+        METADATA_SERVER_EMBEDDED_APPLIED_LSN,
+        Unit::Count,
+        "Raft Metadata raft applied lsn"
+    );
+
+    describe_gauge!(
+        METADATA_SERVER_EMBEDDED_COMMITTED_LSN,
+        Unit::Count,
+        "Raft Metadata raft committed lsn"
+    );
+
+    describe_gauge!(
+        METADATA_SERVER_EMBEDDED_FIRST_INDEX,
+        Unit::Count,
+        "Raft Metadata raft first index"
+    );
+    describe_gauge!(
+        METADATA_SERVER_EMBEDDED_LAST_INDEX,
+        Unit::Count,
+        "Raft Metadata raft last index"
```
embedded vs. raft vs. replicated? cc. @tillrohrmann

tillrohrmann replied:
I've renamed the metadata server kind to "replicated" in this PR: https://github.com/restatedev/restate/pull/2632.

AhmedSoliman commented on crates/metadata-server/src/raft/network/connection_manager.rs:
```diff
@@ -122,6 +127,11 @@ where
                         Some(message) => {
                             match message {
                                 Ok(mut message) => {
+                                    counter!(METADATA_SERVER_EMBEDDED_RECV_MESSAGE_TOTAL, "peer" => self.remote_peer.to_string())
+                                        .increment(1);
+                                    counter!(METADATA_SERVER_EMBEDDED_RECV_MESSAGE_BYTES, "peer" => self.remote_peer.to_string())
+                                        .increment(message.payload.len() as u64);
```
Let's remove the peer address, this is literally infinite cardinality as those sockets get random ports

muhamadazmy replied:
This only holds the plain node id. But if it doesn't add any value i can remove it.

AhmedSoliman replied:
If it's not the address, then it's okay.

AhmedSoliman commented on crates/metadata-server/src/raft/network/connection_manager.rs:
```diff
@@ -181,18 +191,32 @@ impl<M> ConnectionManagerInner<M> {
+        // it's just queued. The actual message might still not get
+        // sent
+
+        counter!(METADATA_SERVER_EMBEDDED_SENT_MESSAGE_TOTAL, "peer" => self.remote_peer.to_string())
+            .increment(1);
+        counter!(METADATA_SERVER_EMBEDDED_SENT_MESSAGE_BYTES, "peer" => self.remote_peer.to_string())
+            .increment(size as u64);
```
ditto

muhamadazmy commented:
@AhmedSoliman thank you for your helpful input. I processed all your comments. Feel free to take another look, leave comments or let me know if it's Approved :)

AhmedSoliman reviewed (approved):
Looks ready to go
